src/ApiCore/GrpcTransport.php

```diff
@@ -0,0 +1,242 @@
+            return ['authorization' => ['Bearer ' . $token['access_token']]];
+        };
+
+        $stubOpts = [
+            'force_new' => $options['forceNewChannel']
+        ];
+        // We need to use array_key_exists here because null is a valid value
+        if (!array_key_exists('sslCreds', $options)) {
+            $stubOpts['credentials'] = $this->createSslChannelCredentials();
+        } else {
+            $stubOpts['credentials'] = $options['sslCreds'];
+        }
+
+        parent::__construct(
+            $host,
+            $stubOpts,
+            $this->pluck('channel', $options, false)
+        );
+    }
+
+    /**
+     * @param Call $call
+     * @param CallSettings $settings The call settings to use for this call.
+     * @param string $streamingType
+     * @param string $resourcesGetMethod
+     *
+     * @return StreamingCallInterface
+     * @todo interface for streaming calls?
+     * @todo pass along resourceGetMethod
+     */
+    public function startStreamingCall(
+        Call $call,
+        CallSettings $settings,
+        $streamingType,
+        $resourcesGetMethod = null
+    ) {
+        $this->validateStreamingApiCallSettings($settings);
+
+        switch ($streamingType) {
+            case 'ClientStreaming':
+                return new ClientStream(
+                    $this->_clientStreamRequest(
+                        $call->getMethod(),
+                        $call->getDecodeType(),
+                        $settings->getUserHeaders() ?: [],
+                        $this->getOptions($settings)
+                    )
+                );
+            case 'ServerStreaming':
+                $message = $call->getMessage();
+
+                if (!$message) {
+                    throw new \Exception('$message is required for ServerStreaming calls.');
+                }
+
+                return new ServerStream(
+                    $this->_serverStreamRequest(
+                        $call->getMethod(),
+                        $message,
+                        $call->getDecodeType(),
+                        $settings->getUserHeaders() ?: [],
+                        $this->getOptions($settings)
+                    )
+                );
+            case 'BidiStreaming':
+                return new BidiStream(
+                    $this->_bidiRequest(
+                        $call->getMethod(),
+                        $call->getDecodeType(),
+                        $settings->getUserHeaders() ?: [],
+                        $this->getOptions($settings)
+                    )
+                );
+            default:
+                throw new ValidationException("Unexpected gRPC streaming type: $streamingType");
+        }
+    }
+
+    private function getOptions(CallSettings $settings)
+    {
+        $retrySettings = $settings->getRetrySettings();
+        $options = $settings->getGrpcOptions() ?: []
+            + ['call_credentials_callback' => $this->credentialsCallback];
```

Comment:
Add brackets around `$settings->getGrpcOptions() ?: []` for clarity, even if that is not strictly necessary. Otherwise, it is unclear whether this is always adding `[]+ ['call_credentials_callback' => $this->credentialsCallback]`, or adding `['call_credentials_callback' => $this->credentialsCallback]` to the result of the expression.

Reply:
All set